File: src/main/java/komu/blunt/parser/TypeParser.java
package komu.blunt.parser;

import komu.blunt.types.*;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static com.google.common.base.Preconditions.checkNotNull;
import static komu.blunt.parser.TokenType.*;
import static komu.blunt.types.Type.*;

public final class TypeParser {

    private final Lexer lexer;

    @SuppressWarnings("unchecked")
    private static final List<TokenType<?>> START_TOKENS =
            Arrays.asList(LPAREN, LBRACKET, IDENTIFIER, TYPE_OR_CTOR_NAME);

    public static Type parseType(String s) {
        return new TypeParser(new Lexer(s)).parseType();
    }

    public static Scheme parseScheme(String s) {
        Qualified<Type> type = parseQualified(s);
        return Qualified.quantifyAll(type);
    }
    
    public static Qualified<Type> parseQualified(String s) {
        TypeParser parser = new TypeParser(new Lexer(s));
        return parser.parseQualified();
    }

    private Qualified<Type> parseQualified() {
        List<Predicate> predicates = parseOptionalPredicates();
        Type type = parseType();
        return new Qualified<Type>(predicates, type);
    }

    private List<Predicate> parseOptionalPredicates() {
        LexerState lexerState = lexer.save();
        try {
            List<Predicate> predicates = new ArrayList<Predicate>();
            if (lexer.readMatchingToken(LPAREN)) {
                do {
                    predicates.add(parsePredicate());
                } while (lexer.readMatchingToken(COMMA));
                lexer.expectToken(RPAREN);
            } else {
                predicates.add(parsePredicate());
            }
            lexer.expectToken(TokenType.BIG_RIGHT_ARROW);
            return predicates;
        } catch (SyntaxException e) {
            lexer.restore(lexerState);
            return Collections.emptyList();
        }
    }

    private Predicate parsePredicate() {
        String className = lexer.readTokenValue(TYPE_OR_CTOR_NAME);
        Type type = parseType();
        return Predicate.isIn(className, type);
    }

    TypeParser(Lexer lexer) {
        this.lexer = checkNotNull(lexer);
    }
    
    public Type parseType() {
        Type type = parseBasic();
        
        while (lexer.readMatchingToken(RIGHT_ARROW))
            type = functionType(type, parseType());

        return type;
    }

    private Type parseBasic() {
        if (lexer.nextTokenIs(TYPE_OR_CTOR_NAME))
            return parseTypeConcrete();
        else
            return parseTypePrimitive();
    }

    public Type parseTypePrimitive() {
        if (lexer.nextTokenIs(LPAREN))
            return parseParens();
        else if (lexer.nextTokenIs(LBRACKET))
            return parseBrackets();
        else if (lexer.nextTokenIs(IDENTIFIER))
            return parseTypeVariable();
        else if (lexer.nextTokenIs(TYPE_OR_CTOR_NAME))
            return genericType(lexer.readTokenValue(TYPE_OR_CTOR_NAME));
        else
            throw lexer.expectFailure("type");
    }

    public Type parseTypeConcrete() {
        String name = lexer.readTokenValue(TYPE_OR_CTOR_NAME);
        
        List<Type> args = new ArrayList<Type>();
        while (lexer.nextTokenIsOneOf(START_TOKENS))
            args.add(parseTypePrimitive());
        
        return genericType(name, args);
    }

    private Type parseParens() {
        lexer.expectToken(LPAREN);

        if (lexer.readMatchingToken(RPAREN))
            return Type.UNIT;

        List<Type> types = new ArrayList<Type>();
        types.add(parseType());
        
        while (lexer.readMatchingToken(COMMA))
            types.add(parseType());
        
        lexer.expectToken(RPAREN);

        if (types.size() == 1)
            return types.get(0);
        else
            return tupleType(types);
    }

    private Type parseBrackets() {
        lexer.expectToken(LBRACKET);
        Type elementType = parseType();
        lexer.expectToken(RBRACKET);

        return listType(elementType);
    }
    
    public TypeVariable parseTypeVariable() {
        String name = lexer.readTokenValue(IDENTIFIER);
        return typeVariable(name);
    }
}

File: src/main/kotlin/komu/blunt/parser/TypeParser.kt
package komu.blunt.parser

import komu.blunt.parser.TokenType.Companion.BIG_RIGHT_ARROW
import komu.blunt.parser.TokenType.Companion.IDENTIFIER
import komu.blunt.parser.TokenType.Companion.LBRACKET
import komu.blunt.parser.TokenType.Companion.LPAREN
import komu.blunt.parser.TokenType.Companion.RIGHT_ARROW
import komu.blunt.parser.TokenType.Companion.RPAREN
import komu.blunt.parser.TokenType.Companion.TYPE_OR_CTOR_NAME
import komu.blunt.types.*
import java.util.*
import java.util.Collections.emptyList

class TypeParser(val lexer: Lexer) {

    constructor(s: String): this(Lexer(s)) { }

    companion object {
        private val START_TOKENS =
            listOf(LPAREN, LBRACKET, IDENTIFIER, TYPE_OR_CTOR_NAME)

        fun parseType(s: String): Type =
            TypeParser(Lexer(s)).parseType()

        fun parseScheme(s: String): Scheme =
            parseQualified(s).quantifyAll()

        fun parseQualified(s: String): Qualified<Type> =
            TypeParser(s).parseQualified()
    }

    private fun parseQualified(): Qualified<Type> {
        val predicates = parseOptionalPredicates()
        val type = parseType()
        return Qualified(predicates, type)
    }

    private fun parseOptionalPredicates(): List<Predicate> {
        if (hasPredicate()) {
            val predicates = if (lexer.peekTokenType() == LPAREN)
                lexer.inParens { lexer.commaSep { parsePredicate() } }
            else
                listOf(parsePredicate())

            lexer.expectToken(BIG_RIGHT_ARROW)

            return predicates

        } else {
            return emptyList()
        }
    }

    private fun hasPredicate(): Boolean =
        lexer.withSavedState {
            try {
                lexer.readMatchingToken(LPAREN)
                parsePredicate()
                true
            } catch (e: SyntaxException) {
                false
            }
        }

    private fun parsePredicate(): Predicate {
        val className = lexer.readTokenValue(TYPE_OR_CTOR_NAME)
        val type = parseType()
        return isIn(className, type)
    }

    fun parseType(): Type {
        var type = parseBasic()

        while (lexer.readMatchingToken(RIGHT_ARROW))
            type = Type.function(type, parseType())

        return type
    }

    private fun parseBasic(): Type =
        if (lexer.nextTokenIs(TYPE_OR_CTOR_NAME))
            parseTypeConcrete()
        else
            parseTypePrimitive()

    fun parseTypePrimitive(): Type =
        when {
            lexer.nextTokenIs(LPAREN) ->
                parseParens()
            lexer.nextTokenIs(LBRACKET) ->
                parseBrackets()
            lexer.nextTokenIs(IDENTIFIER) ->
                parseTypeVariable()
            lexer.nextTokenIs(TYPE_OR_CTOR_NAME) ->
                Type.generic(lexer.readTokenValue(TYPE_OR_CTOR_NAME))
            else ->
                lexer.expectFailure("type")
        }

    fun parseTypeConcrete(): Type {
        val name = lexer.readTokenValue(TYPE_OR_CTOR_NAME)

        val args = ArrayList<Type>()
        while (lexer.nextTokenIsOneOf(START_TOKENS))
            args.add(parseTypePrimitive())

        return Type.generic(name, args)
    }

    private fun parseParens(): Type =
        lexer.inParens {
            if (lexer.nextTokenIs(RPAREN))
                BasicType.UNIT
            else
                Type.tupleOrSingle(lexer.commaSep { parseType() })
        }

    private fun parseBrackets(): Type =
        Type.list(lexer.inBrackets { parseType() })

    fun parseTypeVariable(): Type.Var =
        Type.Var(lexer.readTokenValue(IDENTIFIER))
}

